pkg/cmd/roachtest/roachtestutil/mixedversion/mixedversion.go

```diff
@@ -895,9 +899,143 @@ func (t *Test) plan() (plan *TestPlan, retErr error) {
+	for i := 0; i < len(allVersions)-1; i++ {
+		v := allVersions[i]
+		from := v.String()
+		// startHooks are only run when the cluster is starting up. They are not subject to minimumSupportedVersion.
+		if i == 0 {
+			for _, h := range startHooks {
+				if userHooks[from] == nil {
+					userHooks[from] = make(map[string]bool)
+				}
+				// N.B. combine name and id for debugging purposes; i.e., yields more descriptive error messages.
+				userHooks[from][h.name+"_"+h.id] = true
+				startHooksFound++
+			}
+		}
+		// N.B. user hooks are invoked only from versions that are at least minimumSupportedVersion.
+		if !v.AtLeast(t.options.minimumSupportedVersion) {
+			continue
+		}
+		// firstVersionHooks are only run once we reach an upgrade which satisfies minimumSupportedVersion.
+		if firstVersionHooksFound == 0 {
+			for _, h := range firstVersionHooks {
+				if userHooks[from] == nil {
+					userHooks[from] = make(map[string]bool)
+				}
+				userHooks[from][h.name+"_"+h.id] = true
+				firstVersionHooksFound++
+			}
+		}
+		for _, h := range otherVersionHooks {
+			if userHooks[from] == nil {
+				userHooks[from] = make(map[string]bool)
+			}
+			userHooks[from][h.name+"_"+h.id] = true
+			otherVersionHooksFound++
+		}
+	}
+
+	if assertStartHooks && startHooksFound == 0 {
+		return errors.Newf("expected at least one start hook, but none found\n%s", plan.prettyPrintInternal(true))
+	}
+	if assertFirstVersionHooks && firstVersionHooksFound == 0 {
+		return errors.Newf("expected at least one hook in the first upgrade, but none found\n%s", plan.prettyPrintInternal(true))
+	}
+	if assertOtherVersionHooks && otherVersionHooksFound == 0 {
+		return errors.Newf("expected at least one hook in other upgrades, but none found\n%s", plan.prettyPrintInternal(true))
+	}
+	// Visit each step and remove the corresponding	user hook.
+	plan.iterateSingleSteps(func(ss *singleStep, _ bool) []testStep {
+		from := ss.context.FromVersion().String()
+		if hook, ok := ss.impl.(runHookStep); ok {
+			delete(userHooks[from], hook.hook.name+"_"+hook.hook.id)
+		}
+		return nil
+	})
+	// Check that no unused hooks remain.
+	for from := range userHooks {
+		for hookName := range userHooks[from] {
+			return errors.Newf("unused user hook: %q in version: %q\n%s", hookName, from, plan.prettyPrintInternal(true))
+		}
+	}
+	return nil
+}
+
+// genId generates a unique ID for the hook. The ID is of the form filename:line{_i}, where filename:line
+// corresponds to where in the user code, the hook is created.
+//
+// Since filename is relative, i.e., not fully qualified package path, it may collide with another user hook. In that
+// case, we add a suffix _i, where i is the insertion position of the hook.
+// Note, it's assumed genId is called immediately after the corresponding hook is created; otherwise, filename:line
+// will denote an _internal_ caller.
+func genId(t *Test) string {
```

Comment:
Was very helpful in conjunction with `assertExpectedUserHooks`; methinks it should facilitate debugging user hooks, so why not keep it.

Reply:
I like it! I do wonder if (for aesthetic reasons), it should be hidden via the `debug` flag in the plan output though. It just seems like outside of test eng's eyes it would only be noise. Maybe `MVT_DRY_RUN=true` also enables debug for easy reproduction?

Reply:
>  It just seems like outside of test eng's eyes it would only be noise. 

I was secretly hoping it would improve the observability of user hooks; i.e., when we get a new GH issue, we would immediately see what hooks were used, if any. Of course, that's a lazy way of doing it; we can hide it behind `debug` flag, and perhaps some other, more structured way of tracking user hook stats. I'll create a follow-up issue for it.

Reply:
> I do wonder if (for aesthetic reasons), it should be hidden via the debug flag in the plan output though.

Yep, it was getting too verbose for some of the plans. I've changed the signature of `Description(debug bool)` which for now is only used for an extended description of user hooks. In future, we could sprinkle more debug info for other steps.